File: python/arxiv_download.py
import os
import re
import requests
import xml.etree.ElementTree as ET
import argparse
from tqdm import tqdm
import configparser
from bs4 import BeautifulSoup

def sanitize_filename(filename):
    filename = re.sub(r'[\/:*?"<>|\n]', "-", filename)
    filename = re.sub(r'\s+', " ", filename)
    return filename.strip()

def get_paper_title_from_arxiv(arxiv_id, proxies=None):
    """Retrieve paper title from arXiv using the arXiv API."""
    print("Retrieving paper title from arXiv by ID \033[92m{}\033[0m...".format(arxiv_id))
    url = f"http://export.arxiv.org/api/query?id_list={arxiv_id}"
    response = requests.get(url, proxies=proxies)
    response.raise_for_status()  # Ensure we got a valid response
    
    # Parse the XML response
    root = ET.fromstring(response.content)
    # Find the title within the 'entry' element
    title_element = root.find(".//{http://www.w3.org/2005/Atom}entry/{http://www.w3.org/2005/Atom}title")
    if title_element is None:
        print("\033[91mThe arXiv ID {} is not valid, please confirm...\033[0m".format(arxiv_id))
        return None
    
    title = title_element.text.strip()
    print("Paper title retrieved:\n\t\033[92m<{}>\033[0m".format(title))
    return title

def download_pdf(arxiv_link=None, download_folder=None, proxies=None, filename_pattern="{arxiv_id} - {title}"):
    # Set default download folder to the user's Downloads directory if none is provided
    if not download_folder:
        download_folder = os.path.join(os.path.expanduser("~"), "Downloads")
    
    arxiv_id = arxiv_link.split('/')[-1]  # Extract the arXiv ID from the link
    pdf_url = f"https://arxiv.org/pdf/{arxiv_id}.pdf"
    
    # Retrieve title for the filename from arXiv
    title = get_paper_title_from_arxiv(arxiv_id, proxies=proxies)
    if title is None:
        return
    clean_title = sanitize_filename(title)
    
    # Combine arXiv ID and title to form the filename
    file_name = filename_pattern.format(arxiv_id=arxiv_id, title=clean_title)
    
    # Full path for saving the file
    file_path = os.path.join(download_folder, f"{file_name}.pdf")

    try:
        # Download the PDF file
        print("Downloading PDF...")
        response = requests.get(pdf_url, stream=True, proxies=proxies)
        response.raise_for_status()  # Check if the request was successful

        # Get the total file size
        total_size = int(response.headers.get('content-length', 0))
        block_size = 1024  # 1 Kibibyte

        # Save the PDF to the specified file path with a progress bar
        with open(file_path, 'wb') as pdf_file, tqdm(
            total=total_size, unit='iB', unit_scale=True
        ) as progress_bar:
            for data in response.iter_content(block_size):
                progress_bar.update(len(data))
                pdf_file.write(data)

        print(f"PDF downloaded successfully: {file_path}")
    except requests.exceptions.RequestException as e:
        print(f"Failed to download PDF: {e}")


def extract_arxiv_ids_from_url(weburl):
    # Send a GET request to fetch the page content
    response = requests.get(weburl)

    # Check if the request was successful
    if response.status_code == 200:
        # Parse the HTML content of the page using BeautifulSoup
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Define a regular expression pattern for matching Arxiv IDs
        arxiv_pattern = re.compile(r'https?://arxiv\.org/\w+/(\d+\.\d+)', re.IGNORECASE)
        
        # Search for all links that match the Arxiv URL pattern
        arxiv_ids = set(re.findall(arxiv_pattern, soup.get_text()))

        if arxiv_ids:
            arxiv_ids_str = ', '.join(arxiv_ids)
            print(f"arXiv ID(s) found in the page: \033[92m{arxiv_ids_str}\033[0m")
            return arxiv_ids_str
        else:
            print("\033[91mNo arXiv ID found in the page.\033[0m")
            return ""
    else:
        print(f"\033[91mFailed to fetch page: {response.status_code}\033[0m")
        return ""



def main():
    parser = argparse.ArgumentParser(description="Download PDF from arXiv link(s) or ID(s) or web URL that contains arXiv link, use comma to separate multiple IDs")
    parser.add_argument("-a", "--arxiv", required=True, help="arXiv IDs (comma separated for multiple IDs), arxiv links, any web URL that contains arxiv links")
    parser.add_argument("-f", "--folder", default=None, help="Folder to save the downloaded PDF (default: Downloads folder)")


    args = parser.parse_args()

    # Load configurations from config file
    config = configparser.ConfigParser()
    config_path = os.path.join(os.path.dirname(__file__), 'arxiv.cfg')
    config.read(config_path)

    # Proxy settings
    proxies = {
        'http': config.get('Proxy', 'http', fallback=None),
        'https': config.get('Proxy', 'https', fallback=None)
    }
    
    # Default download folder
    if args.folder is None:
        args.folder = config.get('Settings', 'default_download_folder', fallback=os.path.join(os.path.expanduser("~"), "Downloads"))
    
    # Check if the arXiv argument is a arxiv link or a web URL
    if "arxiv.org" in args.arxiv:
        # Extract arXiv ID from the arXiv link
        arxiv_id = args.arxiv.split('/')[-1]
        args.arxiv = arxiv_id
    elif args.arxiv.startswith("http"):
        # Extract arXiv IDs from the web URL
        arxiv_ids = extract_arxiv_ids_from_url(args.arxiv)
        if not arxiv_ids:
            return
        args.arxiv = arxiv_ids
    else: # the argument must be a comma-separated list of arXiv IDs
        pass

    
    # Filename pattern
    filename_pattern = config.get('Settings', 'filename_pattern', fallback="{arxiv_id} - {title}")
    
    # Split the input arXiv IDs by comma or space
    arxiv_ids = re.split(r'[,\s]+', args.arxiv.strip())

    # Call download_pdf for each arXiv ID
    for arxiv_id in arxiv_ids:
        if len(arxiv_id) != 1:
            print(f"\033[93m[{arxiv_ids.index(arxiv_id)+1}/{len(arxiv_ids)}]\033[0m Downloading PDF for arXiv ID: \033[92m{arxiv_id}\033[0m")
        download_pdf(arxiv_link=f"https://arxiv.org/abs/{arxiv_id}", download_folder=args.folder, proxies=proxies, filename_pattern=filename_pattern)

if __name__ == "__main__":
    main()

File: chrome_ext/arxiv_download.js
function sanitizeFilename(filename) {
    return filename.replace(/[\/:*?"<>|\n]/g, "-").replace(/\s+/g, " ").trim();
}

async function getPaperTitleFromArxiv(arxivId) {
    console.log(`Retrieving paper title from arXiv by ID ${arxivId}...`);
    const url = `https://export.arxiv.org/api/query?id_list=${arxivId}`;
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        xhr.open('GET', url, true);
        xhr.onload = function() {
            if (xhr.status === 200) {
                try {
                    const parser = new DOMParser();
                    const xmlDoc = parser.parseFromString(xhr.responseText, "text/xml");
                    const titleElement = xmlDoc.querySelector("entry > title");
                    if (!titleElement) {
                        console.error(`The arXiv ID ${arxivId} is not valid, please confirm...`);
                        resolve(null);
                        return;
                    }
                    const title = titleElement.textContent.trim();
                    console.log(`Paper title retrieved:\n\t<${title}>`);
                    resolve(title);
                } catch (error) {
                    console.error(`Error parsing response for arXiv ID ${arxivId}:`, error);
                    resolve(null);
                }
            } else {
                console.error(`The arXiv ID ${arxivId} is not valid, HTTP status: ${xhr.status}`);
                resolve(null);
            }
        };
        xhr.onerror = function() {
            console.error(`Network error while fetching arXiv ID ${arxivId}`);
            resolve(null);
        };
        xhr.send();
    });
}

async function downloadPdf(arxivLink, downloadFolder, filenamePattern = "{arxiv_id} - {title}") {
    const arxivId = arxivLink.split('/').pop();
    console.log('arxivId:', arxivId);
    
    const pdfUrl = `https://arxiv.org/pdf/${arxivId}.pdf`;
    console.log('pdfUrl:', pdfUrl);
    
    const title = await getPaperTitleFromArxiv(arxivId);
    console.log('title:', title);
    
    if (!title) return;
    const cleanTitle = sanitizeFilename(title);
    const fileName = filenamePattern
        .replace("{arxiv_id}", arxivId)
        .replace("{title}", cleanTitle);

    try {
        console.log("Initiating PDF download...");
        if (!chrome.downloads) {
            throw new Error("Chrome downloads API not available. Check extension permissions.");
        }
        
        console.log(`Using filename pattern: ${filenamePattern}`);
        console.log(`Final filename: ${fileName}.pdf`);
        
        chrome.downloads.download({
            url: pdfUrl,
            filename: `${fileName}.pdf`,
            saveAs: false
        }, (downloadId) => {
            if (chrome.runtime.lastError) {
                console.error("Download failed:", chrome.runtime.lastError);
                alert(`Download failed: ${chrome.runtime.lastError.message}`);
                return;
            }
            console.log(`PDF download started with ID: ${downloadId}`);
        });
    } catch (error) {
        console.error("Error during download:", error);
        alert(`Error during download: ${error.message}`);
    }
}

async function extractArxivIds(input) {
    // If input is a arxiv.org link
    if (input.includes("arxiv.org") && /\d+\.\d+(?:v\d+)?/.test(input)) {
        const arxivId = input.match(/\d+\.\d+(?:v\d+)?/)[0];
        return [arxivId];
    }

    // If input is a direct arXiv ID
    const matches = input.match(/\d+\.\d+(?:v\d+)?/);
    if (matches) {
        return [matches[0]];
    }
    
    // If input is a URL (including local files)
    if (input.startsWith("http") || input.startsWith("file:///")) {
        // For non-arXiv URLs, fetch and parse content
        try {
            return await fetchFromActiveTab();
        } catch (error) {
            console.error("Error fetching URL:", error);
            return [];
        }
    }
    
    return [];
}


async function fetchFromActiveTab() {
    return new Promise(async (resolve) => {
        try {
            const [tab] = await chrome.tabs.query({ active: true, currentWindow: true });
            console.log('Current tab:', tab);
            // Check if current page is a PDF
            const isPdf = tab.url.toLowerCase().endsWith('.pdf') || 
                         tab.url.toLowerCase().includes('arxiv.org/pdf/') ||
                         (tab.contentType && tab.contentType.toLowerCase() === 'application/pdf');
            console.log('Processing tab:', {
                url: tab.url,
                isPdf: isPdf,
                contentType: tab.contentType || 'unknown'
            });
            
            if (isPdf) {
                console.log('Trying to extract from PDF');
                const results = await chrome.scripting.executeScript({
                    target: { tabId: tab.id },
                    func: async () => {
                        console.log('Attempting to extract text content from PDF viewer...');
                        
                        // Define the getPdfContent function within the execution context
                        async function getPdfContent() {
                            const pdfViewer = document.querySelector('.textLayer');
                            if (pdfViewer) {
                                return pdfViewer.textContent;
                            }
                            
                            const alternativeSelectors = [
                                '#viewer', 
                                '.pdfViewer', 
                                '#viewerContainer' 
                            ];
                            
                            for (const selector of alternativeSelectors) {
                                const element = document.querySelector(selector);
                                if (element) {
                                    return element.textContent;
                                }
                            }
                            return null;
                        }

                        // Call the function and return the content
                        const content = await getPdfContent();
                        console.log('PDF content retrieved:', {
                            found: !!content,
                            length: content?.length,
                            preview: content?.substring(0, 200)
                        });
                        // inline arxiv pattern processing function
                        const arxivPattern = /(?:arxiv\.org\/(?:\w+)\/|arxiv:[\s]*)(\d+\.\d+(?:v\d+)?)/gi;
                        const matches = [...content.matchAll(arxivPattern)];
                        return [...new Set(matches.map(match => match[1]))];
                    }
                });
                resolve(results[0].result || []);
                return;
            }
            
            // For regular web pages
            // Log tab info before processing
            console.log('Processing regular webpage:', {
                url: tab.url,
                title: tab.title
            });
            const results = await chrome.scripting.executeScript({
                target: { tabId: tab.id },
                func: () => {
                    console.log('Executing script to parse arXiv IDs from content...');
                    const content = document.documentElement.innerHTML;
                    // inline arxiv pattern processing function
                    const arxivPattern = /(?:arxiv\.org\/(?:\w+)\/|arxiv:[\s]*)(\d+\.\d+(?:v\d+)?)/gi;
                    const matches = [...content.matchAll(arxivPattern)];
                    return [...new Set(matches.map(match => match[1]))];
                }
            });
            resolve(results[0].result || []);
        } catch (error) {
            console.error('Error executing script:', error);
            resolve([]);
        }
    });
}

async function main(arxiv, folder = null, customFilenamePattern = null) {
    console.log('Input arxiv:', arxiv);

    const config = {
        settings: {
            defaultDownloadFolder: "Downloads",
            filenamePattern: customFilenamePattern || "{arxiv_id} - {title}"
        }
    };

    console.log('Config:', config);

    const downloadFolder = folder || config.settings.defaultDownloadFolder;
    const filenamePattern = customFilenamePattern || "{arxiv_id} - {title}";

    let arxivIds;
    if (arxiv.includes("arxiv.org")) {
        const matches = arxiv.match(/(\d+\.\d+(?:v\d+)?)/);
        console.log('URL matches:', matches);
        arxivIds = matches ? [matches[1]] : [];
    } else if (arxiv.startsWith("http")) {
        const extractedIds = await extractArxivIds(arxiv);
        console.log('Extracted IDs:', extractedIds);
        arxivIds = extractedIds;
    } else {
        arxivIds = arxiv.split(/[, ]+/).filter(id => /^\d+\.\d+(?:v\d+)?$/.test(id));
    }

    console.log('Final arxivIds:', arxivIds);

    if (arxivIds.length === 0) {
        console.error("No valid arXiv IDs found");
        return;
    }

    for (const arxivId of arxivIds) {
        console.log(`Downloading PDF for arXiv ID: ${arxivId}`);
        console.log(`Using filename pattern: ${filenamePattern}`);
        await downloadPdf(`https://arxiv.org/abs/${arxivId}`, downloadFolder, filenamePattern);
    }
}